### ISSUE TITLE
install: Support reinstallation

### ISSUE THREAD

**Issue by cgwalters** [containers/bootc#404]
https://github.com/containers/bootc/pull/137 is tracking part of this, a lot of related discussion in https://github.com/coreos/fedora-coreos-tracker/issues/399

Basically we should support something like `bootc reinstall` that would configurably wipe local system state.

On the implementation side...I think the best way to do this would be something like:

- Create new deployment with fresh `etc` targeting a new stateroot
- Boot into it
- optional: clean up old deployment + stateroot

The beauty of this approach is that the "factory reset" would actually also be fully transactional, i.e. you could roll back into the previous state.  It'd be easy to copy over any desired state (SSH host keys, cached container images e.g.) from the previous boot before deleting it.

**Comment by cgwalters** [comment]
BTW one tricky thing in this is kernel arguments...cc https://github.com/containers/bootc/pull/401 

And also related to this, we should probably go to some effort to preserve the original kargs.

---

An *entirely* different way to implement this would be to retain a pinned version of the original installed deployment.  Some people will want to do that, and we already effectively support it with `ostree admin pin`.  However...rolling back to that in the general case will imply rolling back to something pretty outdated, which we don't want to encourage.  So I think we should focus more on better capturing and maintaining the bits of state that we want from that which *mostly* I think boils down to kargs.

**Comment by vrothberg** [comment]
I'd love to have such a reset functionality for `bootc-{switch,update}` as well.

**Comment by kfox1111** [comment]
Maybe an option to reset state on just a single file in etc or a path under etc too, rather then a complete reset.

**Comment by Malix-Labs** [comment]
Hello folks, is there a clear path of what to do next to make this happen?

**Comment by cgwalters** [comment]
So https://github.com/containers/bootc/pull/137 landed some important preparatory code here. However...that's covering the "reinitialize system from a *different* bootc container image". What we obviously could streamline is the flow of "reinitialize system from the booted image". 

Probably something like `bootc install to-existing-root --from-booted`? And we need to bikeshed what state preservation options we support.

**Comment by imbev** [comment]
How should `/home` and `/etc/passwd` be handled?

**Comment by cgwalters** [comment]
Opened https://github.com/bootc-dev/bootc/pull/1389 for this

**Comment by cgwalters** [comment]
> How should /home and /etc/passwd be handled?

I think in the core design (as used by #1389) it's ~trivial to copy over whatever state one wants from the previous root. But it'd definitely be nice to add some sugar for this.

**Comment by cgwalters** [comment]
xref https://www.freedesktop.org/software/systemd/man/devel/systemd-factory-reset.html - some integration work to do here

**Comment by mrguitar** [comment]
> But it'd definitely be nice to add some sugar for this.

I'm trying to think of areas users may hurt themselves with this. I'm not sure we can add sugar for everything, but I agree w/ @imbev that users & /home dirs are a good starting point. Other contenders could be fstab or mount unints, host ssh keys (probably not that important), Networking (static IPs or bonds), hostname? I wish we had a clean way to reapply whatever anaconda did and/or cloud-init. In lieu of that, maybe some command flags to preserve some things items will help w/ foot guns. $0.02

**Comment by imbev** [comment]
> > But it'd definitely be nice to add some sugar for this.
> 
> I'm trying to think of areas users may hurt themselves with this. I'm not sure we can add sugar for everything, but I agree w/ [@imbev](https://github.com/imbev) that users & /home dirs are a good starting point. Other contenders could be fstab or mount unints, host ssh keys (probably not that important), Networking (static IPs or bonds), hostname? I wish we had a clean way to reapply whatever anaconda did and/or cloud-init. In lieu of that, maybe some command flags to preserve some things items will help w/ foot guns. $0.02

Anaconda could be used with the `/root/anaconda-ks.cfg` generated by the initial installation, but that is an additional dependency.

I agree about the value of the factory reset having flags and file config options for each part.

**Comment by cgwalters** [comment]
Note that in the current design we will always default to a *nondestructive* factory reset by creating a new stateroot. Then users are required to clean up the prior stateroot when they're happy. If it doesn't work, then a rollback into the previous deployment undoes everything.

**Comment by mrguitar** [comment]
Yeah I get that. I think we both want to avoid users resetting their systems, rebooting, and not being able to log in. :) Is the sugar you were thinking of doing the copy for the user? Or were you thinking it could work differently?

**Comment by cgwalters** [comment]
My instinct here is to land some experimental form of this ASAP and gather feedback from ~real world experience.

**Comment by cgwalters** [comment]
> Anaconda could be used with the /root/anaconda-ks.cfg generated by the initial installation, but that is an additional dependency.

Yes! We had a team chat around this issue and for sure it'd be awesome to have Anaconda be useful in this setup to do a reinstallation into a new stateroot.

In theory we could have something like factory reset plugins `/usr/lib/bootc/factory-reset.d` or something that could help guide users who did an install with anaconda to also do a factory reset that way? Or well, a `fedora-derivative` feature flag that 

bootc-image-builder/osbuild are not really designed for this, but maybe could be reworked to do so.

**Comment by cgwalters** [comment]
Ideally we minimize "magic" here; admins should know where the data is coming from and why it's preserved.

**Comment by cgwalters** [comment]
As far as "being able to log in", one pattern we use for https://github.com/bootc-dev/bcvk (and is also used by podman-bootc) is https://github.com/bootc-dev/bcvk/blob/aae33b4a252fea309bd38bfd3dba992187656b22/crates/kit/src/sshcred.rs#L16 where we use [systemd credentials](https://systemd.io/CREDENTIALS/) to add a tmpfiles.d unit that injects a SSH key via SMBIOS (virt level creds). Notably, factory reset at the bootc level won't get rid of that key.

**Comment by ScatteredCognition** [comment]
> How should `/home` and `/etc/passwd` be handled?

I'm not a dev, just a hobbyist, but I'd like to provide my thoughts regarding the matter.
Atleast on Silverblue with OStree, when I attempted a "Factory Reset" (see [this](https://gitlab.com/fedora/ostree/sig/-/issues/28#note_2784899511)) the GNOME OOTB/firstboot wizard appeared just like during the first boot after a fresh install and all the relevant files were auto-generated.

Instead of copying/migrating files like `/etc/passwd`, bootc or another program could provide a firstboot/OOTB experience.
I was thinking something like a systemd service like `bootc-firstboot.service`, which checks for a file named `/etc/bootc-firstboot-complete` (could be another file) and if it doesn't exist, prompts the user to configure their locale, keyboard layout and setup users just like the GNOME OOTB/firstboot wizard, or maybe even prompt the user for a "setup file" that contains all that information.

**Comment by cgwalters** [comment]
> I was thinking something like a systemd service like bootc-firstboot.service, 

What would it provide over `systemd-firstboot` and gnome-initial-setup (and other tools like Ignition or cloud-init)?

Now one issue right now is the Fedora bootc base image has this https://gitlab.com/fedora/bootc/base-images/-/blame/main/minimal/manifest.yaml#L18 but that's effectively a legacy bug. Though changing it would require analysis of consumers.

**Comment by ScatteredCognition** [comment]
> > I was thinking something like a systemd service like bootc-firstboot.service,
> 
> What would it provide over `systemd-firstboot` and gnome-initial-setup (and other tools like Ignition or cloud-init)?

`systemd-firstboot` is exactly what I was describing. I thought something like this didn't exist as other spins like Kinoite and such couldn't survive `--no-merge` "Factory Reset" deployments.

**Comment by cgwalters** [comment]
I think `bootc install factory-reset --print-etc-diff` might be a good replacement for `ostree admin config-diff`
